File: cmux-tui/bindings/java/src/com/cmux/Ids.java
package com.cmux;

import java.util.Objects;

/** Opaque, resource-specific protocol-v2 identifiers. */
public final class Ids {
    public sealed interface Id permits MachineId, SessionId, WorkspaceId, ScreenId,
            PaneId, TabId, TerminalId, BrowserId, ConnectedClientId, SplitId,
            NotificationId, AgentId, StreamId, ProjectionId, PairingRequestId,
            SidebarViewId {
        String value();
    }

    public record MachineId(String value) implements Id {
        public MachineId { value = validate(value, "machine"); }
    }
    public record SessionId(String value) implements Id {
        public SessionId { value = validate(value, "session"); }
    }
    public record WorkspaceId(String value) implements Id {
        public WorkspaceId { value = validate(value, "ws"); }
    }
    public record ScreenId(String value) implements Id {
        public ScreenId { value = validate(value, "screen"); }
    }
    public record PaneId(String value) implements Id {
        public PaneId { value = validate(value, "pane"); }
    }
    public record TabId(String value) implements Id {
        public TabId { value = validate(value, "tab"); }
    }
    public record TerminalId(String value) implements Id {
        public TerminalId { value = validate(value, "term"); }
    }
    public record BrowserId(String value) implements Id {
        public BrowserId { value = validate(value, "browser"); }
    }
    public record ConnectedClientId(String value) implements Id {
        public ConnectedClientId { value = validate(value, "client"); }
    }
    public record SplitId(String value) implements Id {
        public SplitId { value = validate(value, "split"); }
    }
    public record NotificationId(String value) implements Id {
        public NotificationId { value = validate(value, "notification"); }
    }
    public record AgentId(String value) implements Id {
        public AgentId { value = validate(value, "agent"); }
    }
    public record StreamId(String value) implements Id {
        public StreamId { value = validate(value, "stream"); }
    }
    public record ProjectionId(String value) implements Id {
        public ProjectionId { value = validate(value, "projection"); }
    }
    public record PairingRequestId(String value) implements Id {
        public PairingRequestId { value = validate(value, "pairing"); }
    }
    public record SidebarViewId(String value) implements Id {
        public SidebarViewId { value = validate(value, "sidebar_view"); }
    }
    private Ids() {}

    private static String validate(String value, String prefix) {
        Objects.requireNonNull(value, "value");
        String marker = prefix + "_";
        if (!value.startsWith(marker) || value.length() != marker.length() + 32) {
            throw new IllegalArgumentException(
                "expected " + marker + " followed by 32 lowercase hexadecimal characters"
            );
        }
        for (int index = marker.length(); index < value.length(); index++) {
            char character = value.charAt(index);
            if (!((character >= '0' && character <= '9') ||
                    (character >= 'a' && character <= 'f'))) {
                throw new IllegalArgumentException(
                    "expected " + marker + " followed by 32 lowercase hexadecimal characters"
                );
            }
        }
        return value;
    }
}

File: cmux-tui/bindings/python/cmux/ids.py
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Generic, Literal, Type, TypeVar, Union


class ResourceId(str):
    """Validated opaque public resource ID."""

    PREFIX = ""

    def __new__(cls, value: str) -> "ResourceId":
        if not isinstance(value, str):
            raise TypeError(f"{cls.__name__} must be a string")
        expected = rf"{re.escape(cls.PREFIX)}_[0-9a-f]{{32}}"
        if re.fullmatch(expected, value) is None:
            raise ValueError(
                f"{cls.__name__} must be {cls.PREFIX}_ followed by "
                "32 lowercase hexadecimal characters"
            )
        return str.__new__(cls, value)


class MachineId(ResourceId):
    PREFIX = "machine"


class SessionId(ResourceId):
    PREFIX = "session"


class WorkspaceId(ResourceId):
    PREFIX = "ws"


class ScreenId(ResourceId):
    PREFIX = "screen"


class PaneId(ResourceId):
    PREFIX = "pane"


class TabId(ResourceId):
    PREFIX = "tab"


class TerminalId(ResourceId):
    PREFIX = "term"


class BrowserId(ResourceId):
    PREFIX = "browser"


class ConnectedClientId(ResourceId):
    PREFIX = "client"


class SplitId(ResourceId):
    PREFIX = "split"


class StreamId(ResourceId):
    PREFIX = "stream"


class NotificationId(ResourceId):
    PREFIX = "notification"


class AgentId(ResourceId):
    PREFIX = "agent"


class ProjectionId(ResourceId):
    PREFIX = "projection"


class PairingRequestId(ResourceId):
    PREFIX = "pairing"


class SidebarViewId(ResourceId):
    PREFIX = "sidebar_view"


IdT = TypeVar("IdT", bound=ResourceId)
SelectorKind = Literal["id", "current", "name"]


@dataclass(frozen=True)
class Selector(Generic[IdT]):
    """A tagged typed ID, current marker, or exact resource name."""

    kind: SelectorKind
    value: Union[IdT, str, None] = None

    @classmethod
    def by_id(cls, value: IdT) -> "Selector[IdT]":
        if not isinstance(value, ResourceId):
            raise TypeError("ID selectors require a typed resource ID")
        return cls("id", value)

    @classmethod
    def current(cls) -> "Selector[IdT]":
        return cls("current")

    @classmethod
    def name(cls, value: str) -> "Selector[IdT]":
        if not isinstance(value, str):
            raise TypeError("selector name must be a string")
        return cls("name", value)

    def encode(self) -> str:
        if self.kind == "id" and isinstance(self.value, ResourceId):
            return str(self.value)
        if self.kind == "current" and self.value is None:
            return "current"
        if self.kind == "name" and isinstance(self.value, str):
            return f"name:{self.value}"
        raise ValueError("invalid selector")


SelectorInput = Union[Selector[IdT], IdT]


def encode_selector(value: SelectorInput[IdT], expected: Type[IdT]) -> str:
    if isinstance(value, expected):
        return str(value)
    if isinstance(value, Selector):
        if value.kind == "id" and not isinstance(value.value, expected):
            raise TypeError(f"selector requires {expected.__name__}")
        return value.encode()
    raise TypeError(f"selector requires {expected.__name__} or Selector")


__all__ = [
    "AgentId",
    "BrowserId",
    "ConnectedClientId",
    "MachineId",
    "NotificationId",
    "PairingRequestId",
    "PaneId",
    "ProjectionId",
    "ResourceId",
    "ScreenId",
    "Selector",
    "SelectorInput",
    "SessionId",
    "SidebarViewId",
    "SplitId",
    "StreamId",
    "TabId",
    "TerminalId",
    "WorkspaceId",
]
